File: circom/scripts/gen_regex.py
import os
import subprocess
import re


def to_pascal_case(text):
    """Converts a snake_case or kebab-case string to PascalCase."""
    # Replace hyphens with underscores, then split by underscores
    s = re.sub(r"[-_]+", "_", text)
    return "".join(word.capitalize() for word in s.split("_"))


def main():
    # Corrected project_root calculation
    # __file__ is circom/scripts/gen_regex.py
    # os.path.dirname(__file__) is circom/scripts/
    # os.path.join(os.path.dirname(__file__), "..", "..") goes up two levels
    project_root = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))

    # Adjust these paths if your script's location is different
    # relative to the zk-regex project root
    compiler_executable = os.path.join(project_root, "target", "release", "zk-regex")
    regex_dir = os.path.join(project_root, "circom", "regexes")
    output_dir = os.path.join(project_root, "circom", "circuits", "common")
    proving_framework = "circom"

    if not os.path.exists(compiler_executable):
        print(f"Error: Compiler executable not found at {compiler_executable}")
        print(
            "Please build the compiler first (e.g., cargo build in the compiler directory)."
        )
        return

    if not os.path.isdir(regex_dir):
        print(f"Error: Regex directory not found at {regex_dir}")
        return

    os.makedirs(output_dir, exist_ok=True)
    print(f"Ensured output directory exists: {output_dir}")

    print(f"Scanning for JSON files in: {regex_dir}")
    for filename in os.listdir(regex_dir):
        if filename.endswith(".json"):
            json_file_path = os.path.join(regex_dir, filename)
            base_name = os.path.splitext(filename)[0]

            # Convert base_name to PascalCase and append "Regex"
            # e.g., "email_address" -> "EmailAddressRegex"
            # e.g., "simple" -> "SimpleRegex"
            template_name = to_pascal_case(base_name)

            print(f"\nProcessing {filename}...")
            print(f"  Input JSON: {json_file_path}")
            print(f"  Template Name: {template_name}")
            print(f"  Output Directory: {output_dir}")

            command = [
                compiler_executable,
                "decomposed",
                "-d",
                json_file_path,
                "-o",
                output_dir,
                "-t",
                template_name,
                "-p",
                proving_framework,
            ]

            try:
                print(f"  Executing: {' '.join(command)}")
                process = subprocess.run(
                    command, capture_output=True, text=True, check=True
                )
                print("  Compiler Output:")
                if process.stdout:
                    for line in process.stdout.strip().split("\n"):
                        print(f"    STDOUT: {line}")
                if process.stderr:
                    for line in process.stderr.strip().split("\n"):
                        print(
                            f"    STDERR: {line}"
                        )  # Compiler warnings might go to stderr
                print(f"  Successfully generated files for {template_name}")

            except subprocess.CalledProcessError as e:
                print(f"  Error processing {filename}:")
                print(f"    Return code: {e.returncode}")
                if e.stdout:
                    print("    Stdout:")
                    for line in e.stdout.strip().split("\n"):
                        print(f"      {line}")
                if e.stderr:
                    print("    Stderr:")
                    for line in e.stderr.strip().split("\n"):
                        print(f"      {line}")
            except Exception as e:
                print(f"  An unexpected error occurred with {filename}: {e}")

    print("\nScript finished.")


if __name__ == "__main__":
    main()

File: circom/scripts/gen-regex.ts
#!/usr/bin/env node

/**
 * TypeScript version of gen_regex.py
 * Generates Circom regex circuits from JSON definitions
 */

import * as path from 'path';
import { fileURLToPath } from 'url';
import { 
  logger, 
  toPascalCase, 
  ensureDirectory, 
  listFilesWithExtension,
  fileExists,
  executeCargo,
  ScriptError 
} from '../../scripts/utils/index.js';

// Get current file directory (equivalent to __file__ in Python)
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

/**
 * Main function that processes all JSON files and generates Circom circuits
 */
async function main(): Promise<void> {
  try {
    // Calculate project root (accounting for dist directory if compiled)
    // If running from dist/circom/scripts/, go up to zk-regex root
    // If running from circom/scripts/, go up to zk-regex root
    const projectRoot = __dirname.includes('/dist/') 
      ? path.resolve(__dirname, '..', '..', '..', '..') 
      : path.resolve(__dirname, '..', '..');
    
    // Define paths
    const compilerExecutable = path.join(projectRoot, 'target', 'release', 'zk-regex');
    const regexDir = path.join(projectRoot, 'circom', 'regexes');
    const outputDir = path.join(projectRoot, 'circom', 'circuits', 'common');
    const provingFramework = 'circom';

    logger.info(`Project root: ${projectRoot}`);

    // Validate compiler executable exists
    if (!(await fileExists(compilerExecutable))) {
      throw new ScriptError(
        `Compiler executable not found at ${compilerExecutable}. ` +
        'Please build the compiler first (e.g., cargo build in the compiler directory).'
      );
    }

    // Validate regex directory exists
    if (!(await fileExists(regexDir))) {
      throw new ScriptError(`Regex directory not found at ${regexDir}`);
    }

    // Ensure output directory exists
    await ensureDirectory(outputDir);
    logger.info(`Ensured output directory exists: ${outputDir}`);

    // Get all JSON files in regex directory
    logger.info(`Scanning for JSON files in: ${regexDir}`);
    const jsonFiles = await listFilesWithExtension(regexDir, '.json');

    if (jsonFiles.length === 0) {
      logger.warn('No JSON files found in regex directory');
      return;
    }

    // Process each JSON file
    for (const filename of jsonFiles) {
      const jsonFilePath = path.join(regexDir, filename);
      const baseName = path.parse(filename).name;

      // Convert base_name to PascalCase
      const templateName = toPascalCase(baseName);

      logger.info(`\nProcessing ${filename}...`);
      logger.info(`  Input JSON: ${jsonFilePath}`);
      logger.info(`  Template Name: ${templateName}`);
      logger.info(`  Output Directory: ${outputDir}`);

      // Execute the compiler
      const result = executeCargo('run', [
        '--bin', 'zk-regex',
        'decomposed',
        '-d', jsonFilePath,
        '-o', outputDir,
        '-t', templateName,
        '-p', provingFramework,
      ], {
        cwd: projectRoot,
        showOutput: true,
      });

      if (result.success) {
        logger.info(`  Successfully generated files for ${templateName}`);
      } else {
        logger.error(`  Error processing ${filename}:`, {
          returnCode: 'Non-zero exit',
          stdout: result.stdout,
          stderr: result.stderr,
        });
        
        // Continue processing other files even if one fails
        continue;
      }
    }

    logger.info('\nScript finished successfully.');

  } catch (error) {
    if (error instanceof ScriptError) {
      logger.error(error.message);
      if (error.cause) {
        logger.debug('Underlying error:', { error: error.cause.message });
      }
    } else {
      logger.error('An unexpected error occurred:', { 
        error: error instanceof Error ? error.message : String(error) 
      });
    }
    process.exit(1);
  }
}

// Run the script if this file is executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((error) => {
    logger.error('Unhandled error in main:', { error: error.message });
    process.exit(1);
  });
}